hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -62,10 +62,16 @@ public UpgradeDowngrade(
     this.upgradeDowngradeHelper = upgradeDowngradeHelper;
   }
 
-  public boolean needsUpgradeOrDowngrade(HoodieTableVersion toVersion) {
-    HoodieTableVersion fromVersion = metaClient.getTableConfig().getTableVersion();
-    // Ensure versions are same
-    return toVersion.versionCode() != fromVersion.versionCode();
+  public boolean needsUpgradeOrDowngrade(HoodieTableVersion toWriteVersion) {
+    HoodieTableVersion fromTableVersion = metaClient.getTableConfig().getTableVersion();
+
+    if (!config.autoUpgrade() && fromTableVersion.versionCode() < toWriteVersion.versionCode()) {
```

Comment:
auto upgrade is on by default (current behavior). when its not, and versions mismatch for an upgrade, we error out.